modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4IncrementalRequestHandlingIT.java

```diff
@@ -0,0 +1,429 @@
+            for (var opaqueId : ControlServerRequestPlugin.handlers.keySet()) {
+                if (opaqueId.startsWith(testName)) {
+                    var handler = ControlServerRequestPlugin.handlers.get(opaqueId);
+                    handler.recvChunks.forEach(c -> c.chunk.close());
+                    handler.channel.request().getHttpChannel().close();
+                    ControlServerRequestPlugin.handlers.remove(opaqueId);
+                }
+            }
+        }
+
+        ServerRequestHandler awaitRestChannelAccepted(String opaqueId) throws Exception {
+            assertBusy(() -> assertTrue(ControlServerRequestPlugin.handlers.containsKey(opaqueId)));
+            var handler = ControlServerRequestPlugin.handlers.get(opaqueId);
+            safeAwait(handler.channelAccepted);
+            return handler;
+        }
+    }
+
+    static class ServerRequestHandler implements BaseRestHandler.RequestBodyChunkConsumer {
+        final SubscribableListener<Void> channelAccepted = new SubscribableListener<>();
+        final String opaqueId;
+        final BlockingDeque<Chunk> recvChunks = new LinkedBlockingDeque<>();
+        final Netty4HttpRequestBodyStream stream;
+        RestChannel channel;
+
+        boolean recvLast = false;
+
+        ServerRequestHandler(String opaqueId, Netty4HttpRequestBodyStream stream) {
+            this.opaqueId = opaqueId;
+            this.stream = stream;
+        }
+
+        @Override
+        public void handleChunk(RestChannel channel, ReleasableBytesReference chunk, boolean isLast) {
+            recvChunks.add(new Chunk(chunk, isLast));
+        }
+
+        @Override
+        public void accept(RestChannel channel) throws Exception {
+            this.channel = channel;
+            channelAccepted.onResponse(null);
+        }
+
+        void sendResponse(RestResponse response) {
+            channel.sendResponse(response);
+        }
+
+        void consumeBytes(int bytes) {
+            if (recvLast) {
+                return;
+            }
+            while (bytes > 0) {
+                stream.next();
+                var recvChunk = safePoll(recvChunks);
+                bytes -= recvChunk.chunk.length();
+                recvChunk.chunk.close();
+                if (recvChunk.isLast) {
+                    recvLast = true;
+                    break;
+                }
+            }
+        }
+
+        Future<?> onChannelThread(Callable<?> task) {
+            return this.stream.channel().eventLoop().submit(task);
+        }
+
+        record Chunk(ReleasableBytesReference chunk, boolean isLast) {}
+    }
+
+    // takes full control of rest handler from the outside
+    public static class ControlServerRequestPlugin extends Plugin implements ActionPlugin {
+
+        static final String ROUTE = "/_test/request-stream";
+
+        static final ConcurrentHashMap<String, ServerRequestHandler> handlers = new ConcurrentHashMap<>();
```

Comment:
I expect there'll only be one active handler at once, so I think we could use that to simplify things along the lines of  `org.elasticsearch.rest.ChunkedZipResponseIT.RandomZipResponsePlugin#responseRef`.

Reply:
Suppose to be used by pipelining tests where multiple requests in fly. I havent added test yet.